File: Python/DailyLeetCode/minOperationsTree.py
from typing import Optional, List
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def moves(self, a: List[int], b: List[int]) -> int:
        move = 0
        for i in range(len(a)):
            ele = b[i]
            if a[i] == ele:
                continue
            for j in range(i, len(a)):
                if a[j] == ele:
                    break
            a[j] = a[i]
            move += 1
        return move

    def levelOrder(self, root: TreeNode) -> int:
        q = deque([root])
        ans = 0
        while q:
            n = len(q)
            arr = []
            arr2 = []
            for _ in range(n):
                ele = q.popleft()
                if ele.left:
                    q.append(ele.left)
                if ele.right:
                    q.append(ele.right)
                arr.append(ele.val)
                arr2.append(ele.val)
            arr2.sort()
            ans += self.moves(arr, arr2)
        return ans

    def minimumOperations(self, root: Optional[TreeNode]) -> int:
        return self.levelOrder(root)

File: Java/DailyLeetCode/minOperationsTree.java
class Solution {

    public int moves(int[] a,int[] b){
        int move=0;
        int j,i;
        for(i=0;i<a.length;i++){
            int ele=b[i];
            if(a[i]==ele){
                continue;
            }
            else{
                for(j=i;j<a.length;j++){
                    if(a[j]==ele){
                        break;
                    }
                }
                a[j]=a[i];
                move++;
            }
        }
        return move;
    }

    public int levelOrder(TreeNode root){
        Queue<TreeNode> q=new LinkedList<>();
        q.add(root);
        int ans=0;
        while(!q.isEmpty()){
            int n=q.size();
            int[] arr=new int[n];
            int[] arr2=new int[n];
            int index=0;
            for(int i=0;i<n;i++){
                TreeNode ele=q.poll();
                if(ele.left!=null) {
                    q.add(ele.left);
                }
                if(ele.right!=null) q.add(ele.right);

                arr[index]=ele.val;
                arr2[index++]=ele.val;
            }

            Arrays.sort(arr2);
            ans+=moves(arr,arr2);

        }
        return ans;
    }


    public int minimumOperations(TreeNode root) {
        return levelOrder(root);
    }
}